src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -178,20 +301,499 @@ func (s *commitLogSource) ReadData(
+					filePathPrefix, mostRecentSnapshot.ID, infoReaderBufferSize, decoder)
+				if err != nil {
+					s.log.
+						WithFields(
+							xlog.NewField("namespace", mostRecentSnapshot.ID.Namespace),
+							xlog.NewField("blockStart", mostRecentSnapshot.ID.BlockStart),
+							xlog.NewField("shard", mostRecentSnapshot.ID.Shard),
+							xlog.NewField("index", mostRecentSnapshot.ID.VolumeIndex),
+							xlog.NewField("filepaths", mostRecentSnapshot.AbsoluteFilepaths),
+						).
+						Error("error resolving snapshot time for snapshot file")
+
+					// If we couldn't determine the snapshot time for the snapshot file, then rely on
+					// the defer to fallback to using the block start time.
+					return
+				}
+				mostRecentSnapshot.snapshotTime = mostRecentSnapshotTime
+			}()
+		}
+	}
+
+	return mostRecentSnapshotsByBlockShard
+}
+
+func (s *commitLogSource) minimumMostRecentSnapshotTimeByBlock(
+	shardsTimeRanges result.ShardTimeRanges,
+	blockSize time.Duration,
+	mostRecentSnapshotByBlockShard map[xtime.UnixNano]map[uint32]snapshotMetadata,
+) map[xtime.UnixNano]time.Time {
+	minimumMostRecentSnapshotTimeByBlock := map[xtime.UnixNano]time.Time{}
+	for blockStart, mostRecentSnapshotsByShard := range mostRecentSnapshotByBlockShard {
+
+		var minMostRecentSnapshot time.Time
+		for shard, mostRecentSnapshotForShard := range mostRecentSnapshotsByShard {
+			blockRange := xtime.Range{Start: blockStart.ToTime(), End: blockStart.ToTime().Add(blockSize)}
+			if !shardsTimeRanges[shard].Overlaps(blockRange) {
+				// In order for a minimum most recent snapshot to be valid, it needs to be for a block that
+				// we actually need to bootstrap for that shard. This check ensures that our algorithm doesn't
+				// do any extra work, even if we're bootstrapping different blocks for various shards.
+				continue
+			}
+
+			if mostRecentSnapshotForShard.snapshotTime.Before(minMostRecentSnapshot) || minMostRecentSnapshot.IsZero() {
+				minMostRecentSnapshot = mostRecentSnapshotForShard.snapshotTime
+			}
+		}
+
+		if minMostRecentSnapshot.IsZero() {
+			// If we didn't find a minimum most recent snapshot time for this blockStart, just use the
+			// blockStart as the minimum since we'll need to read the entire commit log in this case.
+			minMostRecentSnapshot = blockStart.ToTime()
+		}
+
+		minimumMostRecentSnapshotTimeByBlock[blockStart] = minMostRecentSnapshot
+	}
+
+	return minimumMostRecentSnapshotTimeByBlock
+}
+
+func (s *commitLogSource) bootstrapShardSnapshots(
+	nsID ident.ID,
+	shard uint32,
+	metadataOnly bool,
+	shardTimeRanges xtime.Ranges,
+	blockSize time.Duration,
+	snapshotFiles fs.FileSetFilesSlice,
+	mostRecentCompleteSnapshotByBlockShard map[xtime.UnixNano]map[uint32]snapshotMetadata,
+) (result.ShardResult, error) {
+	var (
+		shardResult    result.ShardResult
+		allSeriesSoFar *result.Map
+		rangeIter      = shardTimeRanges.Iter()
+		err            error
+	)
+
+	for hasMore := rangeIter.Next(); hasMore; hasMore = rangeIter.Next() {
```

Comment:
Can't this just be `for rangeIter.Next() { ... }`?

Reply:
...yes. I don't know why I did that